File: scala/list/move_knight/move_knight.scala

type KnightPos = Tuple2[Int, Int]

val moveKnight = (p: KnightPos) => List(
	(p._1 + 2, p._2 - 1), (p._1 + 2, p._2 + 1),
	(p._1 - 2, p._2 - 1), (p._1 - 2, p._2 + 1),
	(p._1 + 1, p._2 - 2), (p._1 + 1, p._2 + 2),
	(p._1 - 1, p._2 - 2), (p._1 - 1, p._2 + 2)
).filter { case (x, y) => 1 <= x && x <= 8 && 1 <= y && y <= 8 }

println(moveKnight (6, 2))
println(moveKnight (8, 1))

val in3 = (start: KnightPos) => {
	for {
		first <- moveKnight(start)
		second <- moveKnight(first)
		third <- moveKnight(second)
	} yield third
}

println(in3 (6, 2))

val canReachIn3 = (start: KnightPos, end: KnightPos) => in3(start).contains(end)

println(canReachIn3((6, 2), (6, 1)))
println(canReachIn3((6, 2), (7, 3)))

File: Haskell/monad/list/move_knight/move_knight.hs
import Control.Monad

type KnightPos = (Int, Int)

moveKnight :: KnightPos -> [KnightPos]
moveKnight (c, r) = do
	(c', r') <- [
		(c + 2, r - 1), (c + 2, r + 1),
		(c - 2, r - 1), (c - 2, r + 1),
		(c + 1, r - 2), (c + 1, r + 2),
		(c - 1, r - 2), (c - 1, r + 2)
		]
	guard (c' `elem` [1..8] && r' `elem` [1..8])
	return (c', r')

in3 :: KnightPos -> [KnightPos]
in3 start = do
	first <- moveKnight start
	second <- moveKnight first
	moveKnight second

canReachIn3 :: KnightPos -> KnightPos -> Bool
canReachIn3 start end = end `elem` in3 start

main = do
	putStrLn $ show $ moveKnight (6, 2)
	putStrLn $ show $ moveKnight (8, 1)

	putStrLn $ show $ in3 (6, 2)

	putStrLn $ show $ (6, 2) `canReachIn3` (6, 1)
	putStrLn $ show $ (6, 2) `canReachIn3` (7, 3)
